### PR TITLE
8286876: NMT.test_unaliged_block_address_vm_assert fails if using clang toolchain

### PR DESCRIPTION
**Pull request by jdksjolen**
This fixes a bug by removing some UB which Clang used in order to make a test fail. Specifically, Clang optimized the `is_aligned` call in ` MallocHeader::assert_block_integrity` to `true`, because an unaligned `this` is  UB. This is fixed by making a static method, essentially. The new code liberally uses `reinterpret_cast<uintptr_t>()`, this is a specifically mentioned as legal use on https://en.cppreference.com/w/cpp/language/reinterpret_cast .

**The idea is essentially:** Make sure the pointer looks OK before casting it to a `MallocHeader*`.

I also changed the definition of `malloc_header`:

```c++
// old
    return (MallocHeader*)((char*)memblock - sizeof(MallocHeader));
// new
    return &(((MallocHeader*)memblock)[-1]);
```

The previous definition was UB, because you shouldn't cast from a pointer with smaller alignment to one with larger alignment requirements.

I also made `MallocHeader::print_block_on_error` static, as the `this` argument was always equal to the `bad_address`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8286876](https://bugs.openjdk.org/browse/JDK-8286876): NMT.test_unaliged_block_address_vm_assert fails if using clang toolchain

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [22728f6e](https://git.openjdk.org/jdk/pull/11465/files/22728f6e1fe6dbde704ccfd8222602ce850f7da1)
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to [22728f6e](https://git.openjdk.org/jdk/pull/11465/files/22728f6e1fe6dbde704ccfd8222602ce850f7da1)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11465/head:pull/11465` \
`$ git checkout pull/11465`

Update a local copy of the PR: \
`$ git checkout pull/11465` \
`$ git pull https://git.openjdk.org/jdk pull/11465/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11465`

View PR using the GUI difftool: \
`$ git pr show -t 11465`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11465.diff">https://git.openjdk.org/jdk/pull/11465.diff</a>

</details>

### CODE REVIEW

**Comment by tstuefe** [comment]
Hi Johan,

> This fixes a bug by removing some UB which Clang used in order to make a test fail. Specifically, Clang optimized the `is_aligned` call in ` MallocHeader::assert_block_integrity` to `true`, because an unaligned `this` is UB.

Is it allowed to do that? This surprises me. We use is_aligned(this) in other places too, e.g. in oopStorage. So maybe this needs a more general solution. @kimbarrett?

> This is fixed by making a static method, essentially. The new code liberally uses `reinterpret_cast<uintptr_t>()`, this is a specifically mentioned as legal use on https://en.cppreference.com/w/cpp/language/reinterpret_cast .
> 
> **The idea is essentially:** Make sure the pointer looks OK before casting it to a `MallocHeader*`.

Hm. I kind of like to have these as separate discrete operations. I have to think a bit.

> 
> I also changed the definition of `malloc_header`:
> 
> ```c++
> // old
>     return (MallocHeader*)((char*)memblock - sizeof(MallocHeader));
> // new
>     return &(((MallocHeader*)memblock)[-1]);
> ```
> 

okay

> The previous definition was UB, because you shouldn't cast from a pointer with smaller alignment to one with larger alignment requirements.
> 
> I also made `MallocHeader::print_block_on_error` static, as the `this` argument was always equal to the `bad_address`.

It is, but that looks like a regression. It was not the original idea. `this` points to the start of the block, `bad_address` was supposed to point out the point of corruption, which could also be the footer. And in theory also different parts of the header (e.g. for 32-bit, where you have two header canaries, the canary address). There is even logic in print_block_on_error to hexdump two disjunct areas in case the point of corruption is farther away than what one hex dump would cover.

See the explanation, code, and example in the original PR description: https://github.com/openjdk/jdk/pull/5952. 

I see this regressed with "JDK-8292071: NMT: move MallocHeader to its own header and inline header checks". And I managed to break it myself apparently :-(. But the original variant was much nicer and produced better output.

My preference would be to keep handing in two pointers, header address and corruption point, and to re-instate the old behavior. MallocHeader::check_block_integrity already returns the corruption point. The fix is very simple:

```
--- a/src/hotspot/share/services/mallocHeader.inline.hpp
+++ b/src/hotspot/share/services/mallocHeader.inline.hpp
@@ -55,9 +55,7 @@ inline void MallocHeader::assert_block_integrity() const {
   char msg[256];
   address corruption = NULL;
   if (!check_block_integrity(msg, sizeof(msg), &corruption)) {
-    if (corruption != NULL) {
-      print_block_on_error(tty, (address)this);
-    }
+    print_block_on_error(tty, corruption != nullptr ? corruption : (address)this);
     fatal("NMT corruption: Block at " PTR_FORMAT ": %s", p2i(this), msg);
   }
 }
```

**Comment by tstuefe** [comment]
I opened https://bugs.openjdk.org/browse/JDK-8298003 to track the printout error separately and will file a PR for this shortly. I'd like to fix this quickly before JDK 20 code freeze.

**Comment by jdksjolen** [comment]
>I opened https://bugs.openjdk.org/browse/JDK-8298003 to track the printout error separately and will file a PR for this shortly. I'd like to fix this quickly before JDK 20 code freeze.

I see that you got your reviews, so that's good. I'll fix my version by merging on top of yours and doing any necessary adjustments.

>Is it allowed to do that? This surprises me. We use is_aligned(this) in other places too, e.g. in oopStorage. So maybe this needs a more general solution. @kimbarrett?

All I can find is that C++ talks about objects having an "alignment requirement", I take that to mean that the compiler can assume that alignment for something like `this`.

File: src/hotspot/share/services/mallocHeader.hpp
Comment (tstuefe) [inline src/hotspot/share/services/mallocHeader.hpp:135]: This now cuts out functionality from MallocTracker::print_pointer_information(), which had been using check_block_integrity to make sure a pointer is safe to print. Now it resolves the header right away without checking the pointer.

File: src/hotspot/share/services/mallocHeader.inline.hpp
Comment (tstuefe) [inline src/hotspot/share/services/mallocHeader.inline.hpp:88]: Note that it is not necessary to test the supposed malloc header pointer for validity. You can also do the test on the memblock itself. Since that one must always be malloc-aligned, and the malloc header is 16 bytes, it does not matter on which pointer you do the test.

File: src/hotspot/share/services/mallocHeader.hpp
Comment (jdksjolen) [reply src/hotspot/share/services/mallocHeader.hpp]: Fixed this with a templated private variant that the other functions instantiate.

**Comment by jdksjolen** [comment]
> Is it allowed to do that? This surprises me. We use is_aligned(this) in other places too, e.g. in oopStorage. So maybe this needs a more general solution. @kimbarrett?

Hi, I talked to Kim. He referred me to these links:
- https://mail.openjdk.org/pipermail/hotspot-dev/2022-December/067595.html
- https://github.com/openjdk/jdk/pull/10920

In short: Yes, this optimization is allowed and what we're doing is UB. We should definitely look into this.

File: src/hotspot/share/services/mallocHeader.inline.hpp
Comment (tstuefe) [inline src/hotspot/share/services/mallocHeader.inline.hpp:112]: I usually just const_cast to implement one in terms of the other.

**Review by tstuefe — COMMENTED** [review]
Hi @jdksjolen,

I tried to reproduce it with clang 10 and 12 (have no time to install 14), but no such luck. Your tiny reproducer (nice, btw) generates the correct alignment code even for -O3:

```
  7 0000000000000000 <_ZN1A1fEv>:
  8    0:   40 f6 c7 07             test   dil,0x7
  9    4:   0f 94 c0                sete   al
 10    7:   c3                      ret
```

But your repro output is very clear, I don't doubt it.

I still feel this is a general problem though. We use `is_aligned()` to check alignment. This is not an operation that should be optimized away. 

Clang optimizing alignment checks away should hit not only `is_aligned(this)`, but all `is_aligned(xxx)` calls in hotspot with xxx being a pointer to a data structure with pointer-sized members. 

Hotspot rolls its own memory management in many places. So we use placement new a lot: metaspace, hotspot arenas, java heap. When we do, we could err. If we err, we want to know. So alignment tests should always work. `is_aligned` not working is like a smoke detector not working because fire is not allowed.



----

Your patch is mostly fine, but I would like some small reshuffling and renaming. In the end, I'd like:

1) a  public API that checks a *malloced* pointer (the payload pointer) for validity e.g.:

```
static bool MallocHeader::is_valid_malloc_pointer(const void* payload, char* msg, size_t msglen) {
  // range check > 1K
  // (future RFE?) validity checks if we know the range of valid virtual addresse
  // alignment check
  // (future RFE?) possibly, SafeFetch check
}
```

Notes:
	- we don't need to know anything about MallocHeader at this point. If a payload pointer is correctly malloc-aligned, its MallocHeader will be correctly aligned as well.
	- there is no reason for this API to return a corruption address. And there is no reason to hex-dump the block either since we are sure the address cannot point to a valid block. This is usually a case of something misinterpreted as a pointer, and printing the block at that location is usually just misleading.
		
2) A public API that checks a MallocHeader* for validity, minus the pointer tests. We already have this: your variant of MallocHeader::check_block_integrity().

4) A compound operation like your new `assert_block_integrity()`, that does all checks, asserts, and resolves the header. I like clearly named functions though and call it something like `malloc_header_with_checks` or `resolve_checked` (since it now does more than just asserting):

```
static MallocHeader* MallocHeader::resolve_checked(void* payload) {
	if (!MallocHeader::is_valid_malloc_pointer(payload, msg, sizeof(msg)) {
	  fatal("NMT corruption: Block at " PTR_FORMAT ": %s", header, msg);
	}
	MallocHeader* hdr = MemTracker::malloc_header(payload);
	if (!p->check_block_integrity(msg, sizeof(msg), &corruption)) {
		// print block, fatal out
	}
	return hdr;
}
```

This is almost what you already have, but a bit more versatile:
- Majority of cases we use your new compound API to resolve a payload pointer to a header, with all checks and nice error printouts.
- Where we really want to do checks ourselves - eg in debug.cpp, in pp() - we have the building blocks to do so (is_valid_malloc_pointer and check_block_integrity). We can then do our own error reporting.

What do you think?

Cheers Thomas

File: src/hotspot/share/services/mallocHeader.inline.hpp
Comment (tstuefe) [reply src/hotspot/share/services/mallocHeader.inline.hpp:112]: p.s. this was just an observation. Doing it via templates is fine.

**Comment by jdksjolen** [comment]
Thanks for the comments, I implemented the suggested changes. I think we're on the same page, in that the API needed some slight changes but that it was mostly naming that was an issue?

File: src/hotspot/share/services/mallocHeader.cpp
Comment (tstuefe) [inline src/hotspot/share/services/mallocHeader.cpp:36]: Just noticed, this seems wrong. As I wrote before, print_block_on_error() should take two pointers, the header and the corruption point. The header before was implied by this, now that you made it a static method you need to pass the method pointer explicitly. 

That said, what was the reason again this was made static? We only call this now for cases where the header pointer is valid.

File: src/hotspot/share/services/mallocHeader.inline.hpp
Comment (tstuefe) [inline src/hotspot/share/services/mallocHeader.inline.hpp:97]: Please either pass header pointer explicitly, or just leave the original method non-statiic1 and call it like before via header_pointer.

File: src/hotspot/share/services/mallocHeader.inline.hpp
Comment (tstuefe) [inline src/hotspot/share/services/mallocHeader.inline.hpp:92]: Wrong message. Its not a corruption, or at least unlikely. Suggestion: fatal("Not a valid malloc pointer: " ...)

File: src/hotspot/share/services/mallocHeader.cpp
Comment (jdksjolen) [reply src/hotspot/share/services/mallocHeader.cpp:36]: Good job noticing this, I've been staring at this PR for too long :-).

This was a refactoring made because of an old regression, which you fixed in PR #11472. I'll take in your fix by merging with master.

**Comment by jdksjolen** [comment]
@tstuefe , welcome back :). I hope you had a good vacation! Would you mind having a look at this again?

**Review by tstuefe — APPROVED** [review]
Looks good!

**Comment by gerard-ziemski** [comment]
Just FYI I'm taking a look at this...

File: src/hotspot/share/services/mallocHeader.hpp
Comment (albertnetymk) [inline src/hotspot/share/services/mallocHeader.hpp:154]: Is there's a particular reason why in&out types need to match in terms of `const`? Does this work?

```cpp
  inline static MallocHeader* resolve_checked(const void* memblock);
```

Comment (jdksjolen) [reply src/hotspot/share/services/mallocHeader.hpp:154]: Good question. Perhaps in this specific case we can throw away the const information, but it's difficult to discover whether that's true and doesn't look correct to a reader at a first glance. I think an extra method reduces reading complexity such that it is worth keeping.

Comment (tstuefe) [reply src/hotspot/share/services/mallocHeader.hpp:154]: You need both because you should be able to resolve a MallocHeader from a const payload pointer, but that should yield a const pointer too, not a non-const one. Yes, you could return a non-const header pointer, but that would be just bad const discipline.

Comment (albertnetymk) [reply src/hotspot/share/services/mallocHeader.hpp:154]: The input arg should be `const void*`, which accepts both `void*` and `const void*`, because that's a promise to the caller that the current method will not mutate the mem pointed by the input arg. (Ofc, one could throw away `const` and mutate the mem anyway inside the method, but that's just bad code.)

> but that should yield a const pointer too, not a non-const one.

I believe your reasoning is that in&out pointers are essentially the same (off by 1), so they should have the same const-ness.

However, the in&out args are two semantically different entities, `void*` vs `MallocHeader*`. For instance, a caller may construct a `MallocHeader*` out of a `const void*` -- the mem is immutable via `void*` but mutable via `MallocHeader*` due to the extra info offered by the more concrete type. IOW, whether the mem indicated by `MallocHeader*` is mutable or not is up to the caller.

Looking at the `const`-variant caller (I can just find this single one), it's more or less the same as a caller of the non-`const` variant.

```cpp
static void check_expected_malloc_header(const void* payload, MEMFLAGS type, size_t size) {
  const MallocHeader* hdr = MallocHeader::resolve_checked(payload);
  EXPECT_EQ(hdr->size(), size);
  EXPECT_EQ(hdr->flags(), type);
}
```

vs

```cpp
    void* const memblock = ...

    MallocHeader* header2 = MallocHeader::resolve_checked(memblock);
    assert(header2->size() == size, "Wrong size");
    assert(header2->flags() == flags, "Wrong flags");
```

(Seems to me, `header2` should be `const` as well -- here we promise not to mutate the obj via `header2`.)

File: src/hotspot/share/services/mallocTracker.cpp
Comment (tstuefe) [inline src/hotspot/share/services/mallocTracker.cpp:173]: Make const.

File: src/hotspot/share/services/mallocHeader.hpp
Comment (tstuefe) [reply src/hotspot/share/services/mallocHeader.hpp:154]: > The input arg should be `const void*`, which accepts both `void*` and `const void*`, because that's a promise to the caller that the current method will not mutate the mem pointed by the input arg. 

Yes,

(Ofc, one could throw away `const` and mutate the mem anyway inside the method, but that's just bad code.)

Yes.

> 
> > but that should yield a const pointer too, not a non-const one.
> 
> I believe your reasoning is that in&out pointers are essentially the same (off by 1), so they should have the same const-ness.
> 
> However, the in&out args are two semantically different entities, `void*` vs `MallocHeader*`. For instance, a caller may construct a `MallocHeader*` out of a `const void*` -- the mem is immutable via `void*` but mutable via `MallocHeader*` due to the extra info offered by the more concrete type. IOW, whether the mem indicated by `MallocHeader*` is mutable or not is up to the caller.

Not sure if we talk at cross purposes, but:

- const variant is for safe *querying* while preventing accidental write access. You don't need write access for querying.
- non-const variant is for modifying the header. The header is mostly immutable, the only thing we change are the canaries. Without that, we could dispose of the non-const variant altogether.

The whole thing follows the same logic as a class where you differentiate between read-only and write-access for the same accessor, e.g.

```
class X {
   Y y;
   const Y& get_y() const;
   Y& get_y();
}
```

see e.g. Effective C++. Regrettably, C++ forces us to duplicate the method.

> 
> Looking at the `const`-variant caller (I can just find this single one), it's more or less the same as a caller of the non-`const` variant.
> 
> ```c++
> static void check_expected_malloc_header(const void* payload, MEMFLAGS type, size_t size) {
>   const MallocHeader* hdr = MallocHeader::resolve_checked(payload);
>   EXPECT_EQ(hdr->size(), size);
>   EXPECT_EQ(hdr->flags(), type);
> }
> ```
> 
> vs
> 
> ```c++
>     void* const memblock = ...
> 
>     MallocHeader* header2 = MallocHeader::resolve_checked(memblock);
>     assert(header2->size() == size, "Wrong size");
>     assert(header2->flags() == flags, "Wrong flags");
> ```
> 
> (Seems to me, `header2` should be `const` as well -- here we promise not to mutate the obj via `header2`.)

Agreed.

Comment (albertnetymk) [reply src/hotspot/share/services/mallocHeader.hpp:154]: When you agree with "The input arg should be `const void*`...", do you mean this?

```cpp
  inline static const MallocHeader* resolve_checked(const void* memblock);
  inline static       MallocHeader* resolve_checked(const void* memblock);
```
I don't think this is valid C++.

What I suggested is to have a single method and let the caller decide regarding `const` of the returned value.

Comment (tstuefe) [reply src/hotspot/share/services/mallocHeader.hpp:154]: > When you agree with "The input arg should be `const void*`...", do you mean this?
> 
> ```c++
>   inline static const MallocHeader* resolve_checked(const void* memblock);
>   inline static       MallocHeader* resolve_checked(const void* memblock);
> ```
> 
> I don't think this is valid C++.
> 
> What I suggested is to have a single method and let the caller decide regarding `const` of the returned value.

No, I did not mean that. Both functions are needed. What you suggest would lead to a violation of constness. Why drop protection? To gain what, save one line?

Having two functions that differ in constness of input arg (or of this pointer) and output argument is an established practice. The normal recommendation for non-trivial functions then is to implement one in terms of the other by casting constness away. @jdksjolen solved this in a neater, albeit more verbose way with a templatized implementation.

Comment (albertnetymk) [reply src/hotspot/share/services/mallocHeader.hpp:154]: > Both functions are needed.

The signature of `inline static MallocHeader* resolve_checked(void* memblock);` is too permissive -- the callee is allowed to mutate content of `memblock`, but it shouldn't.

> What you suggest would lead to a violation of constness.

As I pointed out before, constructing a `MallocHeader*` from a `const void*` can be a valid use case.

> Why drop protection? To gain what, save one line?

I think the bigger issue is having `void*` as the arg, which forces callers to check the impl of this method.

**Review by gerard-ziemski — APPROVED** [review]
I was a bit concerned over the non-trivial change this fix turned out to be, so I investigated a little.

The smallest fix I can come up with is to change:

```
if (!is_aligned(this, sizeof(uint64_t))) {
```

to

```
  volatile uintptr_t ptr = (uintptr_t)this;
  if (!is_aligned(ptr, sizeof(uint64_t))) {
```

which works, since using volatile turns any compiler optimizations off, so we end up with the needed pointer (? that's how I understand it, but not 100% sure)

Another way I can see this fixed is to pass the `memblock` into `MallocHeader::assert_block_integrity` keeping it as `void*`, which we can then pass right onto `check_block_integrity`, which we need to adapt slightly to receive the pointer.

I am OK with the proposed fix, though I personally would prefer either one of the ones I presented.

**Comment by tstuefe** [comment]
Hi @gerard-ziemski ,

> I was a bit concerned over the non-trivial change this fix turned out to be, so I investigated a little.

thank you for your thoughts. Skimming patches down is always good.
> 
> The smallest fix I can come up with is to change:
> 
> ```
> if (!is_aligned(this, sizeof(uint64_t))) {
> ```
> 
> to
> 
> ```
>   volatile uintptr_t ptr = (uintptr_t)this;
>   if (!is_aligned(ptr, sizeof(uint64_t))) {
> ```

I'd personally be fine with it but the objection would still be that `this` is not allowed to be unaligned.

> 
> which works, since using volatile turns any compiler optimizations off, so we end up with the needed pointer (? that's how I understand it, but not 100% sure)
> 
> Another way I can see this fixed is to pass the `memblock` into `MallocHeader::assert_block_integrity()` keeping it as `void*`, which we can then pass right onto `check_block_integrity()`, which we need to adapt slightly to receive the pointer.
> 

So, make `MallocHeader::assert_block_integrity()` a static function, and pass the void* payload pointer in? That could work.

> I am OK with the proposed fix, though I personally would prefer either one of the ones I presented.

A small patch would be easier to backport and clearer with respect to what is actually the fix.

But Johans patch also has some cleanups I like:
- moving the header resolving functions out of memTracker.hpp into mallocHeader.hpp and making them static members of MallocHeader
- the solution to the cons/nonconst duplication via templates is neat
- -1 instead of sizeof to resolve the header
- and other small things.

So possibly it would be fine to have a minimal patch, and then some cleanups afterwards?

I leave this up to @jdksjolen. He has put a lot of work into it already, and if he wants to be done with it and ship it that totally fine for me.

**Comment by gerard-ziemski** [comment]
> Hi @gerard-ziemski ,
> 
> > I was a bit concerned over the non-trivial change this fix turned out to be, so I investigated a little.
> 
> thank you for your thoughts. Skimming patches down is always good.
> 
> > The smallest fix I can come up with is to change:
> > ```
> > if (!is_aligned(this, sizeof(uint64_t))) {
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to
> > ```
> >   volatile uintptr_t ptr = (uintptr_t)this;
> >   if (!is_aligned(ptr, sizeof(uint64_t))) {
> > ```
> 
> I'd personally be fine with it but still would be UB and other compilers may complain again.
> 
> > which works, since using volatile turns any compiler optimizations off, so we end up with the needed pointer (? that's how I understand it, but not 100% sure)
> > Another way I can see this fixed is to pass the `memblock` into `MallocHeader::assert_block_integrity()` keeping it as `void*`, which we can then pass right onto `check_block_integrity()`, which we need to adapt slightly to receive the pointer.
> 
> So, make `MallocHeader::assert_block_integrity()` a static function, and pass the void* payload pointer in? That could work.
> 
> > I am OK with the proposed fix, though I personally would prefer either one of the ones I presented.
> 
> A small patch would be easier to backport and clearer with respect to what is actually the fix.
> 
> But Johans patch also has some cleanups I like:
> 
> * moving the header resolving functions out of memTracker.hpp into mallocHeader.hpp and making them static members of MallocHeader
> * the solution to the cons/nonconst duplication via templates is neat
> * -1 instead of sizeof to resolve the header
> * and other small things.
> 
> So possibly it would be fine to have a minimal patch, and then some cleanups afterwards?
> 
> I leave this up to @jdksjolen. He has put a lot of work into it already, and if he wants to be done with it and ship it that totally fine for me. I already approved the current form.

Yes, I'm OK with that as well, I just wanted to point out that we could have done this in a simpler way...

**Comment by jdksjolen** [comment]
Hi, thank you for the comments. I'll be merging this as-is when I've confirmed that all tests pass. I'm doing more extensive testing on this PR now.

**Comment by jdksjolen** [comment]
It turns out that IN/OUT are macros in MVSC, so I had to rename those template parameters or the builds will fail on Windows.

**Comment by jdksjolen** [comment]
/integrate

All right, time to merge. Thank you for all the input on this.

**Pull request closed without merging**
